File: ios/FLEXR/Core/Locale/LanguageStore.swift
import Foundation
import Observation

/// Gewählte Sprache, dauerhaft gespeichert.
///
/// Bewusst getrennt vom [SessionStore]: die Sprachwahl gehört dem Gerät, nicht
/// dem Konto — beim Abmelden soll sie stehen bleiben.
///
/// Solange nichts gewählt wurde, gilt [AppLanguage.detect]; ein Wechsel der
/// Systemsprache oder ein Umzug wirkt damit weiter, bis der Nutzer den Regler
/// einmal selbst bedient.
@MainActor
@Observable
final class LanguageStore {

    private static let key = "flexr_language"

    private let defaults: UserDefaults

    /// Ob der Regler auf diesem Gerät je bedient wurde.
    ///
    /// [language] verschweigt den Unterschied bewusst (es liefert dann die
    /// Vorgabe aus [AppLanguage.detect]). Beim Abgleich mit dem Profil zählt er
    /// aber: Eine Wahl auf diesem Gerät schlägt das Profil, eine bloße Vorgabe
    /// nicht.
    private(set) var hasExplicitChoice: Bool

    /// Wird nach jeder Wahl gerufen. Setzt der Einstiegspunkt (`FlexrApp`), um
    /// die Sprache ans Profil zu melden — der Speicher selbst kennt weder
    /// Netzwerk noch Repositories.
    @ObservationIgnored var onChange: ((AppLanguage) -> Void)?

    /// Läuft gerade eine Übernahme vom Profil? Dann ist die Zuweisung an
    /// [language] keine Wahl des Nutzers, und [didSet] darf weder speichern
    /// noch melden. Ohne diese Klammer müsste [adopt] die Nebenwirkungen
    /// hinterher zurücknehmen — das hinge an der Reihenfolge.
    @ObservationIgnored private var isAdopting = false

    /// Die aktuelle Sprache. Änderungen zeichnen die Oberfläche neu — jede View
    /// liest ihre Texte über [strings], und das hängt an dieser Eigenschaft.
    var language: AppLanguage {
        didSet {
            guard language != oldValue else { return }
            // Netzwerkstapel und Hintergrunddienste kennen die Oberfläche
            // nicht und holen ihre Texte aus einer statischen Fassung — die
            // muss mitwandern. Siehe die Begründung bei `FlexrStrings.current`.
            FlexrStrings.current = strings
            guard !isAdopting else { return }
            defaults.set(language.rawValue, forKey: Self.key)
            hasExplicitChoice = true
            onChange?(language)
        }
    }

    /// Texttabelle in der gewählten Sprache.
    var strings: FlexrStrings { FlexrStrings(language: language) }

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        let saved = defaults.string(forKey: Self.key).flatMap(AppLanguage.init(rawValue:))
        hasExplicitChoice = saved != nil
        language = saved ?? AppLanguage.detect()
        FlexrStrings.current = strings
    }

    /// Sprache vom Profil übernehmen, ohne daraus eine Wahl zu machen.
    ///
    /// Für den Fall „neues Gerät, nie gewählt": Die Oberfläche stellt sich um,
    /// [hasExplicitChoice] bleibt aber falsch — erst ein Griff zum Regler macht
    /// daraus eine Entscheidung dieses Geräts.
    func adopt(_ language: AppLanguage) {
        guard !hasExplicitChoice, language != self.language else { return }
        isAdopting = true
        self.language = language
        isAdopting = false
    }
}

File: android-native/app/src/main/java/flexr/social/app/core/locale/LanguageStore.kt
package flexr.social.app.core.locale

import android.content.Context
import android.content.SharedPreferences
import androidx.core.content.edit
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import dagger.hilt.android.qualifiers.ApplicationContext
import flexr.social.app.di.ApplicationScope
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch

/**
 * Bis zum 11.09.2026 lag die Sprachwahl hier in einem DataStore. Sie liegt
 * jetzt in [SharedPreferences] — aus einem einzigen Grund: Sie muss
 * **synchron** lesbar sein.
 *
 * [flexr.social.app.MainActivity.attachBaseContext] setzt die Sprache am
 * Basis-Context der Activity, und dieser Aufruf kann nicht warten: Er
 * entscheidet, welche Ressourcen die Activity ihr ganzes Leben lang liefert,
 * und laeuft ab, bevor Hilt, Compose oder irgendein Coroutine-Scope existiert.
 * DataStore kann das grundsaetzlich nicht bedienen (nur `suspend`), und
 * `runBlocking` auf dem Hauptfaden waere genau das, wovor DataStore warnt.
 * [SharedPreferences] ist fuer diesen Fall gebaut.
 *
 * Die Wahl gehoert weiterhin dem Geraet, nicht dem Konto — beim Abmelden
 * ([flexr.social.app.data.session.SessionStore.clear]) bleibt sie stehen.
 */
private const val PREFERENCES_NAME = "flexr_language"
private const val LANGUAGE_KEY = "app_language"

/**
 * Der alte DataStore. Er existiert nur noch, um die Wahl eines Nutzers, der
 * schon eine aeltere Fassung installiert hatte, einmalig herueberzuholen
 * (siehe [LanguageStore.migrateLegacyChoice]).
 */
private val Context.legacyLanguageDataStore: DataStore<Preferences> by
    preferencesDataStore(name = "flexr_settings")

private val legacyLanguageKey = stringPreferencesKey("app_language")

/**
 * Gewaehlte Sprache, dauerhaft gespeichert.
 *
 * Solange nichts gewaehlt wurde, liefert [language] das Ergebnis von
 * [AppLanguage.detect]; ein Wechsel der Systemsprache oder ein Umzug wirkt
 * damit weiterhin, bis der Nutzer den Regler einmal selbst bedient.
 */
@Singleton
class LanguageStore @Inject constructor(
    @ApplicationContext private val context: Context,
    @ApplicationScope scope: CoroutineScope,
) {

    private val preferences = preferences(context)

    private val state = MutableStateFlow(storedLanguage(context))

    /**
     * Die ausdrueckliche Wahl — `null`, solange der Regler nie bedient wurde.
     *
     * [language] verschweigt den Unterschied bewusst (es liefert dann die
     * Vorgabe aus [AppLanguage.detect]). Beim Abgleich mit dem Profil zaehlt er
     * aber: Eine Wahl auf diesem Geraet schlaegt das Profil, eine blosse
     * Vorgabe nicht.
     */
    val chosen: StateFlow<AppLanguage?> = state.asStateFlow()

    val language: Flow<AppLanguage> = state.map { it ?: AppLanguage.detect() }

    /**
     * Derselbe Wert wie die erste Ausgabe von [language], nur ohne Warten.
     *
     * Sammler benutzen ihn als Startwert ihres `stateIn`, damit niemand fuer
     * einen Frame auf der Vorgabe steht, obwohl die Wahl laengst feststeht —
     * genau dieser Zwischenstand haette in [flexr.social.app.MainActivity]
     * sonst einen ueberfluessigen Neuaufbau ausgeloest.
     */
    val current: AppLanguage get() = state.value ?: AppLanguage.detect()

    init {
        scope.launch { migrateLegacyChoice() }
    }

    suspend fun setLanguage(language: AppLanguage) = store(language)

    private fun store(language: AppLanguage) {
        preferences.edit { putString(LANGUAGE_KEY, language.code) }
        state.value = language
    }

    /**
     * Wahl aus dem alten DataStore uebernehmen, einmalig.
     *
     * Laeuft nur, wenn hier noch nichts steht. Das Ergebnis kommt zu spaet fuer
     * [flexr.social.app.MainActivity.attachBaseContext] dieses einen Starts —
     * die Activity merkt den Unterschied aber selbst und baut sich neu auf.
     */
    private suspend fun migrateLegacyChoice() {
        if (state.value != null) return
        val code = runCatching { context.legacyLanguageDataStore.data.first()[legacyLanguageKey] }
            .getOrNull()
        AppLanguage.fromCode(code)?.let(::store)
    }

    companion object {

        /**
         * Die gespeicherte Wahl, synchron — fuer `attachBaseContext`, das
         * weder `suspend` sein noch auf Hilt warten kann.
         */
        fun storedLanguage(context: Context): AppLanguage? =
            AppLanguage.fromCode(preferences(context).getString(LANGUAGE_KEY, null))

        private fun preferences(context: Context): SharedPreferences =
            (context.applicationContext ?: context)
                .getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE)
    }
}
